q2cli/builtin/tools.py

```diff
@@ -195,7 +195,89 @@ def peek(path):
+              ' specified as either categorical or numeric column types.')
+@click.option('--error-on-extra', is_flag=True,
+              help='Cast flags provided inline that do not exist within the'
+              ' metadata provided will result in a raised error.'
+              ' Enabled by default. If this flag is disabled and'
+              ' extra flags are provided, they will be ignored.')
+@click.option('--ignore-missing', is_flag=True,
+              help='Cast flags not provided inline that exist within the'
+              ' metadata provided will be ignored. Enabled by default.'
+              ' If this flag is disabled and not all column flags'
+              ' are provided, an error will be raised.')
+@click.option('--output-file', required=True,
+              type=click.Path(exists=False, file_okay=True, dir_okay=True,
+                              writable=True),
+              help='Path to file or directory where the'
+              ' modified metdata should be written to.')
+@click.argument('paths', nargs=-1, required=True, metavar='METADATA...',
+                type=click.Path(exists=True, file_okay=True, dir_okay=False,
+                                readable=True))
+def cast_metadata(paths, cast, output_file, error_on_extra,
+                  ignore_missing):
+    import tempfile
+    import qiime2
+
+    metadata = _merge_metadata(paths)
+
+    # TODO: Refactor into separate function
+    # Parsing cast arguments into key:value pairs
+    try:
+        cast_dict = {k: v for k, v in (elem.split(':') for elem in cast)}
+    except Exception as err:
+        header = \
+            'Could not parse provided cast arguments into key:value pairs.'
+        ' Please make sure all castes are of the format --cast COLUMN:TYPE'
+        q2cli.util.exit_with_error(err, header=header)
+
+    column_names = set(metadata.columns.keys())
+    cast_names = set(cast_dict.keys())
+
+    # TODO: Refactor into separate function
+    # Cast flag handling
+    if not error_on_extra:
+        if not cast_names.issubset(column_names):
+            raise click.BadParameter(
+                message='One or more cast columns were not found within the'
+                ' metadata.',
+                param_hint='cast')
+
+    if ignore_missing:
+        if not column_names.issubset(cast_names):
+            raise click.BadParameter(
+                message='One or more columns within the metadata'
+                ' were not provided in the cast.',
+                param_hint='cast')
+
+    # TODO: Refactor into separate function
+    # Casted metadata: writing to a new file
+
+    # Remove entries from the cast dict that are not in the metadata to avoid
+    # errors further down the road
+    for cast in cast_names:
+        if cast not in column_names:
+            cast_dict.pop(cast)
+
+    with tempfile.NamedTemporaryFile() as temp:
+        metadata.save(temp.name)
+        cast_md = qiime2.Metadata.load(temp.name, cast_dict)
```

Comment:
Clever! I forgot about that `.load` option to override the types, cool!

Reply:
🤓 😏

Reply:
The error message raised by invalid castings is pretty hairy:

```bash
qiime tools cast-metadata --cast bar:numeric md1.tsv
Traceback (most recent call last):
  File "/Users/matthew/projects/qiime2/qiime2/qiime2/metadata/io.py", line 109, in read
    df = df.apply(self._cast_column, axis='index',
  File "/Users/matthew/.conda/envs/q2dev/lib/python3.8/site-packages/pandas/core/frame.py", line 7768, in apply
    return op.get_result()
  File "/Users/matthew/.conda/envs/q2dev/lib/python3.8/site-packages/pandas/core/apply.py", line 185, in get_result
    return self.apply_standard()
  File "/Users/matthew/.conda/envs/q2dev/lib/python3.8/site-packages/pandas/core/apply.py", line 276, in apply_standard
    results, res_index = self.apply_series_generator()
  File "/Users/matthew/.conda/envs/q2dev/lib/python3.8/site-packages/pandas/core/apply.py", line 290, in apply_series_generator
    results[i] = self.f(v)
  File "/Users/matthew/.conda/envs/q2dev/lib/python3.8/site-packages/pandas/core/apply.py", line 110, in f
    return func(x, *args, **kwds)
  File "/Users/matthew/projects/qiime2/qiime2/qiime2/metadata/io.py", line 296, in _cast_column
    return self._to_numeric(series)
  File "/Users/matthew/projects/qiime2/qiime2/qiime2/metadata/io.py", line 326, in _to_numeric
    raise MetadataFileError(
qiime2.metadata.io.MetadataFileError: Cannot convert metadata column 'bar' to numeric. The following values could not be interpreted as numeric: 'one', 'three', 'two'

There may be more errors present in the metadata file. To get a full report, sample/feature metadata files can be validated with Keemei: https://keemei.qiime2.org

Find details on QIIME 2 metadata requirements here: https://docs.qiime2.org/2021.8/tutorials/metadata/

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/matthew/.conda/envs/q2dev/bin/qiime", line 33, in <module>
    sys.exit(load_entry_point('q2cli', 'console_scripts', 'qiime')())
  File "/Users/matthew/.conda/envs/q2dev/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/Users/matthew/.conda/envs/q2dev/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/Users/matthew/.conda/envs/q2dev/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/matthew/.conda/envs/q2dev/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/matthew/.conda/envs/q2dev/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/matthew/.conda/envs/q2dev/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/Users/matthew/projects/qiime2/q2cli/q2cli/builtin/tools.py", line 292, in cast_metadata
    cast_md = qiime2.Metadata.load(temp.name, cast_dict)
  File "/Users/matthew/projects/qiime2/qiime2/qiime2/metadata/metadata.py", line 328, in load
    return MetadataReader(filepath).read(into=cls,
  File "/Users/matthew/projects/qiime2/qiime2/qiime2/metadata/io.py", line 120, in read
    raise MetadataFileError(msg, include_suffix=False)
qiime2.metadata.io.MetadataFileError: Cannot convert metadata column 'bar' to numeric. The following values could not be interpreted as numeric: 'one', 'three', 'two'

There may be more errors present in the metadata file. To get a full report, sample/feature metadata files can be validated with Keemei: https://keemei.qiime2.org

Find details on QIIME 2 metadata requirements here: https://docs.qiime2.org/2021.8/tutorials/metadata/
```

One way to clean this up is to wrap the metadata parsing errors:

```python
        try:
            cast_md = qiime2.Metadata.load(temp.name, cast_dict)
        except qiime2.metadata.io.MetadataFileError as e:
            raise click.BadParameter(message=e, param_hint='cast') from e
```

```bash
qiime tools cast-metadata --cast bar:numeric md1.tsv
Usage: qiime tools cast-metadata [OPTIONS] METADATA...

Error: Invalid value for cast: Cannot convert metadata column 'bar' to numeric. The following values could not be interpreted as numeric: 'one', 'three', 'two'

There may be more errors present in the metadata file. To get a full report, sample/feature metadata files can be validated with Keemei: https://keemei.qiime2.org

Find details on QIIME 2 metadata requirements here: https://docs.qiime2.org/2021.8/tutorials/metadata/
```